From 9f86fd1a7abaa8440e9b7c11219a93eb3dd563b5 Mon Sep 17 00:00:00 2001
From: Hariprasad Kelam <hkelam@marvell.com>
Date: Mon, 21 Sep 2020 15:31:35 +0530
Subject: [PATCH 0819/1921] octeontx2-pf: support to change link
 speed/duplex/autoneg

Enable user to change configurations like speed, duplex and
autoneg of network interface via ethtool -s option.

	ethtool -s eth0 advertise 0x1
         --> default values will be passed speed/duplex/an
	ethtool -s eth0 advertise 0x40000 speed 10/1000
                        duplex half/full autoneg on phyad 0x1
	Use phyad 0x1 to overwrite default values

Change-Id: Icdaf8f2bf60801cd81dd293d2ef78a081627b43d
Signed-off-by: Hariprasad Kelam <hkelam@marvell.com>
Reviewed-on: https://sj1git1.cavium.com/c/IP/SW/kernel/linux/+/38069
Tested-by: sa_ip-sw-jenkins <sa_ip-sw-jenkins@marvell.com>
Reviewed-by: Sunil Kovvuri Goutham <Sunil.Goutham@cavium.com>
[WK: The original patch got from Marvell sdk11.21.09]
Signed-off-by: Wenlin Kang <wenlin.kang@windriver.com>
---
 .../marvell/octeontx2/nic/otx2_ethtool.c      | 37 +++++++++++++++++++
 1 file changed, 37 insertions(+)

diff --git a/drivers/net/ethernet/marvell/octeontx2/nic/otx2_ethtool.c b/drivers/net/ethernet/marvell/octeontx2/nic/otx2_ethtool.c
index fbf48d3de2a5..ac07c0e40a6b 100644
--- a/drivers/net/ethernet/marvell/octeontx2/nic/otx2_ethtool.c
+++ b/drivers/net/ethernet/marvell/octeontx2/nic/otx2_ethtool.c
@@ -1198,6 +1198,35 @@ static int otx2_get_link_ksettings(struct net_device *netdev,
 	return 0;
 }
 
+#define OTX2_OVERWRITE_DEF	0x1
+static int otx2_populate_input_params(struct otx2_nic *pfvf,
+				      struct cgx_set_link_mode_req *req,
+				      u32 speed, u8 duplex, u8 autoneg,
+				      u8 phy_address)
+{
+	if (!ethtool_validate_speed(speed) ||
+	    !ethtool_validate_duplex(duplex))
+		return -EINVAL;
+
+	if (autoneg != AUTONEG_ENABLE && autoneg != AUTONEG_DISABLE)
+		return -EINVAL;
+
+	if (phy_address == OTX2_OVERWRITE_DEF) {
+		req->args.speed = speed;
+		/* firmware expects 1 for half duplex and 0 for full duplex
+		 * hence inverting
+		 */
+		req->args.duplex = duplex ^ 0x1;
+		req->args.an = autoneg;
+	} else {
+		req->args.speed = SPEED_UNKNOWN;
+		req->args.duplex = DUPLEX_UNKNOWN;
+		req->args.an = AUTONEG_UNKNOWN;
+	}
+
+	return 0;
+}
+
 static int otx2_set_link_ksettings(struct net_device *netdev,
 				   const struct ethtool_link_ksettings *cmd)
 {
@@ -1222,6 +1251,14 @@ static int otx2_set_link_ksettings(struct net_device *netdev,
 		mutex_unlock(&pfvf->mbox.lock);
 		return -EINVAL;
 	}
+
+	if (otx2_populate_input_params(pfvf, req, cmd->base.speed,
+				       cmd->base.duplex, cmd->base.autoneg,
+				       cmd->base.phy_address)) {
+		mutex_unlock(&pfvf->mbox.lock);
+		return -EINVAL;
+	}
+
 	err =  otx2_sync_mbox_msg(&pfvf->mbox);
 	if (!err) {
 		rsp = (struct cgx_set_link_mode_rsp *)
-- 
2.31.1

